### ISSUE TITLE
[FEATURE-REQUEST] Regex extract and bit shifting

### ISSUE THREAD

**Issue by mansenfranzen** [vaexio/vaex#1154]
Hi Vaex-Community,

I'm fairly new to Vaex and I really like the progress you're making!

I've been using pandas and spark a lot in the past but have come across several performance issues with both recently. I'm tempted to checkout vaex for a current use case but I'm not sure of the required functionality is already present. I went through the documentation and also consulted closed/open issues without success.

### Regex extract
In pandas there is [`extract`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.str.extract.html) and in spark there is [`regexp_extract`](http://spark.apache.org/docs/latest/api/python/pyspark.sql.html?highlight=regex#pyspark.sql.functions.regexp_extract) to extract one or more values from a string via a regular expression. I couldn't find anything comparable to this in vaex. Is there anything related already?

### Bit shifting
In spark there is [`shiftLeft`](http://spark.apache.org/docs/latest/api/python/pyspark.sql.html?highlight=regex#pyspark.sql.functions.shiftLeft) to shift a value by a given number of bits. Is there anything comparable in vaex?

Thanks a lot!

**Comment by maartenbreddels** [comment]
Hi,

thanks!
There are several functions using regex, but the base problem was that we did not have a data structure that would support multiple values, now that we support Apache Arrow, and together with the upstream work at https://github.com/apache/arrow/pull/8459 we will have this in the near future.

All standard Python operators are supported,e.g.:
![image](https://user-images.githubusercontent.com/1765949/105049546-0cf12f00-5a6d-11eb-8424-41882f52b25a.png)

Regards,

Maarten

**Comment by maartenbreddels** [comment]
You can probably use https://vaex.io/docs/api.html#vaex.expression.StringOperations.replace for this (ctrl-f for regex should work)

**Comment by crliu95** [comment]
> Hi,
> 
> thanks!
> There are several functions using regex, but the base problem was that we did not have a data structure that would support multiple values, now that we support Apache Arrow, and together with the upstream work at [apache/arrow#8459](https://github.com/apache/arrow/pull/8459) we will have this in the near future.
> 
> All standard Python operators are supported,e.g.:
> ![image](https://user-images.githubusercontent.com/1765949/105049546-0cf12f00-5a6d-11eb-8424-41882f52b25a.png)
> 
> Regards,
> 
> Maarten

Hi Maarten,

Glad to hear the exciting news that we vaex-users would be able to EFFICIENTLY utilize some functions like `extract` or `extractall` in pandas. That's the very feature I am seeking for weeks!

We have been using vaex to deal with millions of text materials stored in string format, and got struggled to use `re.findall` in the processing. The codes we are using now are like this:
`vaex_df['result'] = '' 
for i in range(0, len(vaex_df)):
    vaex_df['result'][i] = re.findall(re_pattern, vaex_df['text'][i])`
This works but we do not know whether it is efficient (e.g. In pandas we should do something like `apply`).

Anyway, hope the internally supported vaex function could provide a solid and efficient experience. 

Thanks in advance, Maarten!

**Comment by mansenfranzen** [comment]
The above mentioned [apache/arrow#8459](https://github.com/apache/arrow/pull/8459) has been closed and pyarrow 4.0.0 already includes the [extract_regex](https://arrow.apache.org/docs/cpp/compute.html#string-extraction) compute kernel. Will it be included in the next release for vaex, too?

**Comment by mansenfranzen** [comment]
Sure, I'll give it a try. I searched for a developer guide for vaex but I couldn't find any. Can you point me anywhere I could start with? 

From the existing PRs I found #1190 which looks like a good starting point, don't you think?

**Comment by FrankBoermanTenneT** [comment]
hi @maartenbreddels @mansenfranzen I am interested in this feature too. Is there any progress on this?

**Comment by maartenbreddels** [comment]
@mansenfranzen that would indeed be a good start

@FrankBoermanTenneT are you referring to extract_regex?

**Comment by FrankBoermanTenneT** [comment]
@maartenbreddels yes to the regex functionality. I am trying to port some pandas code which includes the df[column].str.extract function so would be nice if that was ported

**Comment by maartenbreddels** [comment]
Note that you can use: https://vaex.io/docs/api.html#vaex.expression.StringOperations.replace as a substitute. But extract would be nicer.

**Comment by FrankBoermanTenneT** [comment]
hi @maartenbreddels how can that be used as a substitute to extract? I am confused

**Comment by maartenbreddels** [comment]
Using capture groups, e.g.:
```
import vaex
text = ['Something', 'very pretty', 'is coming', 'our', 'way.']
df = vaex.from_arrays(text=text)
df.text.str.replace('.*o(.).*', r'\1', regex=True)
```

It's not fully the same though.

**Comment by FrankBoermanTenneT** [comment]
hi @maartenbreddels thanks for the snippit but I find it hard to follow my regex is not that good. Do you mean that you replace everything you dont want and thus extract something? is \1 some kind of negation?

**Comment by maartenbreddels** [comment]
\1 matches what is between the parenthesis, in this case any character after an 'o'

**Comment by FrankBoermanTenneT** [comment]
ah okay, yeah thats gonna be complicated for my usecase. I need to do a pattern like this:
```
A= (?P<A>.*)%;B= (?P<B>.*)%;C= (?P<C>.*)%;B= (?P<B>.*)%
```

**Comment by mansenfranzen** [comment]
@maartenbreddels I will give it a try to provide a PR adding the new regextract functionality from apache arrow. Since the PR mainly wraps apache arrow functionality, it shouldn't be too hard to implement.

However, is there any documentation on how to setup a local development environment and how to run tests for vaex? This would greatly help me getting started. I couldn't find anything related in the docs or elsewhere.

**Comment by maartenbreddels** [comment]
Great @mansenfranzen !

https://vaex.io/docs/installing.html should contain all you need, and maybe it's a good idea to start adding missing parts to that document as well!
for instance, we use pytest, but it's not documented how to use it, and we have some scripts and files in `ci/*` to help set up a new environment. But I think we can do better :)

**Comment by mansenfranzen** [comment]
@maartenbreddels thanks, I wonder why I missed the developer section in the install-docs. It is exacly what I need. 

I was just about to start digging into the CI to see how tests are run. Perhaps I'm going to create a short PR on how to run tests in local dev env along the way.

**Comment by mansenfranzen** [comment]
@maartenbreddels I added a PR #1423 but I'm stuck without further advice from your site regarding some conceptional questions.
